nemo_skills/inference/model/utils.py

```diff
@@ -24,5 +33,54 @@ def trim_after_stop_phrases(text: str, stop_phrases: list[str]) -> str:
     return re.split("|".join(escaped_stop_phrases), text, maxsplit=1)[0]
 
 
+class ServerTokenizer:
+    """Class to encode and decode prompts via POST requests to the tokenizer endpoint."""
+
+    def __init__(self, url):
+        self.tokenizer_url = url
+        self.detokenizer_url = url.replace("/tokenize", "/detokenize")
+
+    def encode(self, prompt: str | list[dict]) -> list[int]:
+        """Encode the prompt using the tokenizer endpoint."""
+        if isinstance(prompt, str):
+            payload = {"prompt": prompt}
+        elif isinstance(prompt, list):
+            payload = {"messages": prompt}
+
+        response = requests.post(self.tokenizer_url, json=payload, timeout=30)
+        response.raise_for_status()
+
+        tokens = response.json()["tokens"]
+        return tokens
+
+    def decode(self, tokens: list) -> str:
+        """Decode a list of tokens using the tokenizer endpoint."""
+        payload = {"tokens": tokens}
+        response = requests.post(self.detokenizer_url, json=payload, timeout=30)
+        response.raise_for_status()
+
+        text = response.json()["prompt"]
+        return text
+
+
+class WrapperAutoTokenizer:
+    """Wrapper around the AutoTokenizer class to provide same interface as the ServerTokenizer class."""
+
+    def __init__(self, model_name: str):
+        LOG.info(f"Initializing tokenizer from string: {model_name}")
+        self.tokenizer = AutoTokenizer.from_pretrained(model_name)
+
+    def encode(self, prompt: str | list[dict]) -> list[int]:
+        """Encode the prompt using the tokenizer."""
+        if isinstance(prompt, str):
+            return self.tokenizer.encode(prompt)
+        elif isinstance(prompt, list):
+            return self.tokenizer.apply_chat_template(prompt, add_generation_prompt=True)
```

Comment:
the `add_generation_prompt=True` is not always correct, maybe add a TODO or something. This function might be called in the middle of the generation (e.g. in TIR mode). But I guess it's not that big a deal as we'd just be subtracting a little more tokens than necessary